### PR TITLE
qemu: receive qmp event when stop vm

### PR DESCRIPTION
**Pull request by Ace-Tang**
receive qmp event when start a qmp connect. This is useful for issue
1368, when send a quit qmp request to qemu, but vm shutdown first, and
not response for the quit qmp request. So receive the shutdown event to
judge whether vm is exited.

Fix: #1368

Signed-off-by: Ace-Tang <aceapril@126.com>

### CODE REVIEW

**Comment by Ace-Tang** [comment]
Hi, @devimc , could you help to see if the patch make sense ?

File: virtcontainers/qemu.go [hunk 1]
```diff
@@ -748,10 +760,20 @@ func (q *qemu) qmpSetup() error {
 	}
 	q.qmpMonitorCh.qmp = qmp
 	q.qmpMonitorCh.disconn = disconnectCh
+	go q.loopQMPEvent(events)
 
 	return nil
 }
 
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
```
Comment (lifupan) [inline virtcontainers/qemu.go]: If the qemu is killed instead of exit by sending SHUTDOWN cmd from qmp, would this loopQMPEvent catch this case?

Comment (Ace-Tang) [reply virtcontainers/qemu.go]: @lifupan , yes, it can be catched

Comment (lifupan) [reply virtcontainers/qemu.go]: Hi @Ace-Tang 
Sorry, I cannot catch when and where the qmp shutdown event was created when qemu was killed?

Comment (Ace-Tang) [reply virtcontainers/qemu.go]: :confused:   @lifupan , Can you describe how you test

Comment (lifupan) [reply virtcontainers/qemu.go]: It's simple, run a pod and then kill the qemu process,  and I didn't see any shutdown events was logged.

Comment (Ace-Tang) [reply virtcontainers/qemu.go]: I am not test with k8s, but only run with ctr run, and kill the qemu, then I can get event
```
time="2019-05-15T00:25:02.069044069+08:00" level=info msg="{\"timestamp\": {\"seconds\": 1557851102, \"microseconds\": 68970}, \"event\": \"SHUTDOWN\", \"data\": {\"guest\": false}}" ID=kill-k1 source=virtcontainers subsystem=qmp
time="2019-05-15T00:25:02.069167555+08:00" level=info msg="log event SHUTDOWN" ID=kill-k1 source=virtcontainers subsystem=qemu
time="2019-05-15T00:25:02.069678935+08:00" level=error msg="Wait for process failed" container=kill-k1 error="rpc error: code = Unavailable desc = transport is closing" pid=kill-k1
```

Comment (jcvenegas) [reply virtcontainers/qemu.go]: @Ace-Tang  can you also get when you do a `kill -9` ?

File: virtcontainers/qemu.go [hunk 2]
```diff
@@ -658,6 +660,14 @@ func (q *qemu) stopSandbox() error {
 
 	err = q.qmpMonitorCh.qmp.ExecuteQuit(q.qmpMonitorCh.ctx)
 	if err != nil {
+		for i := 0; i < 100; i++ {
+			if q.isShutdown {
```
Comment (jcvenegas) [inline virtcontainers/qemu.go]: It is not wrong but looks a bit racy  if this start to be used in other places. 

I wonder if you try to check if  qemu process exist  or the a qemu socket (that is known exist) is a better way to identify if VM is running or not.

Comment (Ace-Tang) [reply virtcontainers/qemu.go]: Make sense, if we do this, we can remove the logic about log qmp event

File: virtcontainers/qemu.go (hunk 1 above)
Comment (Ace-Tang) [reply virtcontainers/qemu.go]: No, so this method is not always work.

**Comment by raravena80** [comment]
@Ace-Tang any updates? Thx!

**Comment by Ace-Tang** [comment]
@raravena80 ， actually, I am not sure it is ok to make this change, code is not elegant

**Comment by raravena80** [comment]
@Ace-Tang do you want to close this? and re-open later with a better change?

File: virtcontainers/qemu.go [hunk 3]
```diff
@@ -658,6 +660,14 @@ func (q *qemu) stopSandbox() error {
 
 	err = q.qmpMonitorCh.qmp.ExecuteQuit(q.qmpMonitorCh.ctx)
```
Comment (amshinde) [inline virtcontainers/qemu.go]: @Ace-Tang I am surprised that the earlier qmp.Setup() succeeds, so looks like for the qemu exits between qmpSetup and  qmp.ExecuteQuit() call.

File: virtcontainers/qemu.go (hunk 2 above)
Comment (amshinde) [reply virtcontainers/qemu.go]: @Ace-Tang I think its better if can check if the qmp.disconnectedCh is closed(indicating conneection has been lost) and check that the qmp.sock no longer exits to verify that the qemu instance has exited on receiving an error on `qmp.ExecuteQuit`. I dont think you can rely on SHUTDOWN msg, esp if you execute a `kill -9` on the qemu process. 

@markdryan What do you think?

**Comment by markdryan** [comment]
> @Ace-Tang I think its better if can check if the qmp.disconnectedCh is closed(indicating conneection has been lost) and check that the qmp.sock no longer exits to verify that the qemu instance has exited on receiving an error on qmp.ExecuteQuit. I dont think you can rely on SHUTDOWN msg, esp if you execute a kill -9 on the qemu process.
> 
> @markdryan What do you think?

Yes this is correct.  Looking at govmm, [this doesn't seem to be documented very well](https://github.com/intel/govmm/blob/master/qemu/qmp.go#L656), which is probably what has led to the confusion.  However, qmp.disconnectedCh, is closed when the instance dies, either of its on accord or because it has been terminated using ExecuteQuit.  There is an example of how to do this.

https://github.com/intel/govmm/blob/master/qemu/examples_test.go#L84

disconnectedCh exists for this very reason.  It provides a race free way for the govmm client to know when a QEMU instance has actually quit.

**Comment by caoruidong** [comment]
ping @Ace-Tang

**Comment by Ace-Tang** [comment]
I will close it now, since I do not have enough time on fix this problem

**Pull request closed without merging**